src/fireedge/src/server/routes/api/2fa/functions.js

```diff
@@ -0,0 +1,291 @@
+    callback: (err, value) => {
+      responseOpennebula(
+        () => undefined,
+        err,
+        value,
+        (info) => {
+          if (info !== undefined && info !== null) {
+            callback(info)
+          } else {
+            next()
+          }
+        },
+        next
+      )
+    },
+  })
+}
+
+/**
+ * Add 2FA user.
+ *
+ * @param {object} res - http response
+ * @param {Function} next - express stepper
+ * @param {object} params - params of http request
+ * @param {string} [params.token] - params of http request
+ * @param {object} userData - user of http request
+ * @param {Function} oneConnection - function of xmlrpc
+ */
+const setup = (
+  res = {},
+  next = defaultEmptyFunction,
+  params = {},
+  userData = {},
+  oneConnection = defaultEmptyFunction
+) => {
+  const { user, password } = userData
+  if (!(user && password)) {
+    next()
+
+    return
+  }
+
+  const { token } = params
+  const oneConnect = oneConnection(user, password)
+  getUserInfoAuthenticated(oneConnect, next, (data) => {
+    const fireedge = data?.USER?.TEMPLATE?.FIREEDGE
+    const secret = fireedge?.[default2FAOpennebulaTmpVar]
+    if (Number.isInteger(parseInt(data?.USER?.ID, 10)) && secret && token) {
+      if (check2Fa(secret, token)) {
+        oneConnect({
+          action: Actions.USER_UPDATE,
+          parameters: [
+            parseInt(data.USER.ID, 10),
+            generateNewResourceTemplate(
+              fireedge || {},
+              { [default2FAOpennebulaVar]: secret },
+              [default2FAOpennebulaTmpVar]
+            ),
+            1,
+          ],
+          callback: (error, value) => {
+            responseOpennebula(
+              () => undefined,
+              error,
+              value,
+              (pass) => {
+                if (pass !== undefined && pass !== null) {
+                  res.locals.httpCode = httpResponse(ok)
+                }
+                next()
+              },
+              next
+            )
+          },
+        })
+      } else {
+        res.locals.httpCode = httpResponse(unauthorized)
+        next()
+      }
+    } else {
+      next()
+    }
+  })
+}
+
+/**
+ * Generate QR.
+ *
+ * @param {object} res - http response
+ * @param {Function} next - express stepper
+ * @param {object} params - params of http request
+ * @param {object} userData - user of http request
+ * @param {Function} oneConnection - function of xmlrpc
+ */
+const qr = (
+  res = {},
+  next = defaultEmptyFunction,
+  params = {},
+  userData = {},
+  oneConnection = defaultEmptyFunction
+) => {
+  const { user, password } = userData
+  if (!(user && password)) {
+    next()
+
+    return
+  }
+
+  const secret = speakeasy.generateSecret({
+    length: 10,
+    name: twoFactorAuthIssuer,
+  })
+  if (secret && secret.otpauth_url && secret.base32) {
+    const { otpauth_url: otpURL, base32 } = secret
+    qrcode.toDataURL(otpURL, (err, dataURL) => {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This route is missing authorization checks. Without proper authorization, sensitive routes may be exposed to unauthorized users, leading to potential data breaches or unauthorized actions. Ensure that authorization middleware or handlers are applied. For more information on missing authorization vulnerabilities, refer to [CWE-862](https://cwe.mitre.org/data/definitions/862.html).